llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp

```diff
@@ -665,6 +665,12 @@ Value *SCEVExpander::visitUDivExpr(const SCEVUDivExpr *S) {
   }
 
   Value *RHS = expand(S->getRHS());
+  if (SafeUDivMode && !SE.isKnownNonZero(S->getRHS())) {
```

Comment:
The logic here doesn't look quite right wrt the poison handling. This code does nothing if isKnownNonZero (but maybe poison). We should at least leave a TODO here.

Reply:
Updated to check to only consider non-zero constants safe, thanks.
Maybe worth introducing a `isKnownNonPoison()` helper?